File: python/dmarebrandsbot/cogs/status.py
from __future__ import annotations

import discord
from discord import app_commands
from discord.ext import commands

from ..formatting import BAD, BRAND, GOOD, WARN, embed, iso_stamp
from ..permissions import requires

TONE = {
    "undetected": GOOD,
    "updating": WARN,
    "offline": WARN,
    "detected": BAD,
    "discontinued": BAD,
}

HEADLINE = {
    "undetected": "Up and safe to use",
    "updating": "Updating",
    "offline": "Temporarily offline",
    "detected": "Detected, tell people to stop",
    "discontinued": "Discontinued",
}


class Status(commands.Cog):
    def __init__(self, bot: commands.Bot) -> None:
        self.bot = bot

    @app_commands.command(
        description="Show whether a product is up, and whether customer keys are frozen"
    )
    @app_commands.describe(product="Limit it to one product, for example rust")
    @requires("status.read")
    async def status(self, interaction: discord.Interaction, product: str | None = None) -> None:
        result = await self.bot.partner_api.status(product)
        rows = result.get("products") or []

        if not rows:
            await interaction.edit_original_response(content="No products came back.")
            return

        worst = next((p for p in rows if p.get("status") != "undetected"), rows[0])
        blocks = []
        for p in rows:
            head = HEADLINE.get(p.get("status"), p.get("status"))
            beta = " · beta" if p.get("beta") else ""
            frozen = "\nKeys are frozen, so nobody is losing time." if p.get("frozen") else ""
            sale = "" if p.get("sellable") else "\nThis one cannot be sold any more."
            note = f"\n{p['message']}" if p.get("message") else ""
            seen = f"\nChanged {iso_stamp(p['updated_at'])}" if p.get("updated_at") else ""
            blocks.append(f"**{p.get('label')}** · {head}{beta}{frozen}{sale}{note}{seen}")

        view = embed("Product status", TONE.get(worst.get("status"), BRAND))
        view.description = "\n\n".join(blocks)
        await interaction.edit_original_response(embed=view)

    @app_commands.command(description="Show the latest Rust build and the recent patch notes")
    @requires("status.read")
    async def updates(self, interaction: discord.Interaction) -> None:
        result = await self.bot.partner_api.updates()
        build = result.get("current_build")
        notes = (result.get("updates") or [])[:5]

        view = embed("Game updates", BRAND)
        if build:
            seen = f" · {iso_stamp(build['at'])}" if build.get("at") else ""
            gap = result.get("average_gap_days")
            extra = f"\nUsually about {gap} days between builds." if gap else ""
            view.description = f"Current build **{build.get('buildid')}**{seen}{extra}"
        else:
            view.description = "No build information yet."

        if notes:
            listed = "\n".join(f"[{n['title']}]({n['url']}) · {n['date']}" for n in notes)
            view.add_field(name="Recent notes", value=listed[:1024])

        await interaction.edit_original_response(embed=view)


async def setup(bot: commands.Bot) -> None:
    await bot.add_cog(Status(bot))

File: node/src/commands/status.js
import { SlashCommandBuilder } from "discord.js";
import { BAD, BRAND, GOOD, WARN, embed, isoStamp } from "../format.js";

const TONE = {
  undetected: GOOD,
  updating: WARN,
  offline: WARN,
  detected: BAD,
  discontinued: BAD,
};

const HEADLINE = {
  undetected: "Up and safe to use",
  updating: "Updating",
  offline: "Temporarily offline",
  detected: "Detected, tell people to stop",
  discontinued: "Discontinued",
};

export const status = {
  data: new SlashCommandBuilder()
    .setName("status")
    .setDescription("Show whether a product is up, and whether customer keys are frozen")
    .addStringOption((o) =>
      o.setName("product").setDescription("Limit it to one product, for example rust")
    ),
  scopes: "status.read",
  async run({ interaction, api }) {
    const wanted = interaction.options.getString("product") ?? undefined;
    const result = await api.status(wanted);
    const rows = result.products ?? [];

    if (!rows.length) {
      await interaction.editReply({ content: "No products came back." });
      return;
    }

    const worst = rows.find((p) => p.status !== "undetected") ?? rows[0];
    const view = embed("Product status", TONE[worst.status] ?? BRAND).setDescription(
      rows
        .map((p) => {
          const head = HEADLINE[p.status] ?? p.status;
          const beta = p.beta ? " · beta" : "";
          const frozen = p.frozen ? "\nKeys are frozen, so nobody is losing time." : "";
          const sale = p.sellable ? "" : "\nThis one cannot be sold any more.";
          const note = p.message ? `\n${p.message}` : "";
          const seen = p.updated_at ? `\nChanged ${isoStamp(p.updated_at)}` : "";
          return `**${p.label}** · ${head}${beta}${frozen}${sale}${note}${seen}`;
        })
        .join("\n\n")
    );

    await interaction.editReply({ embeds: [view] });
  },
};

export const updates = {
  data: new SlashCommandBuilder()
    .setName("updates")
    .setDescription("Show the latest game build and the recent patch notes")
    .addStringOption((o) =>
      o.setName("product").setDescription("Which game, rust or wardogs. Defaults to rust")
    ),
  scopes: "status.read",
  async run({ interaction, api }) {
    const wanted = interaction.options.getString("product") ?? undefined;
    const result = await api.updates(wanted);
    const build = result.current_build;
    const notes = (result.updates ?? []).slice(0, 5);

    const view = embed(`Game updates · ${result.game ?? wanted ?? "rust"}`, BRAND);
    view.setDescription(
      build
        ? `Current build **${build.buildid}**${build.at ? ` · ${isoStamp(build.at)}` : ""}` +
            (result.average_gap_days ? `\nUsually about ${result.average_gap_days} days between builds.` : "")
        : "No build information yet."
    );

    if (notes.length) {
      view.addFields({
        name: "Recent notes",
        value: notes.map((n) => `[${n.title}](${n.url}) · ${n.date}`).join("\n").slice(0, 1024),
      });
    }

    await interaction.editReply({ embeds: [view] });
  },
};
